### PR TITLE
fix: handle missing documents in updateMessageTaskStates and log warn…

### LINKED ISSUE #10944: updateMessageTaskStates crashes when allDocs returns rows without doc
<html><head></head><body><h1><code>updateMessageTaskStates</code> crashes when <code>allDocs</code> returns rows without <code>doc</code></h1>
<!--
**Important**: This is a public repository. Anyone in the world can see what's posted here.
If you are posting screenshots or log files, please **carefully examine them for** the presence
of any kind of **protected health information** (PHI). Images or logs containing PHI _must_ be
posted in fully-redacted form, with no visible PHI.
--!>
<hr>
<h2>Describe the Bug</h2>
<p><code>api/src/services/messaging.js</code> assumes every row returned from <code>db.medic.allDocs({ include_docs: true })</code> contains a <code>doc</code>. That assumption is not always true — CouchDB/PouchDB can omit <code>doc</code> for <code>not_found</code>, deleted, or race-condition rows.</p>
<p>When that happens, <code>updateMessageTaskStates</code> passes <code>undefined</code> entries into task lookup logic and throws a runtime exception:</p>
<pre><code>TypeError: Cannot read properties of undefined (reading 'tasks')
</code></pre>
<p><strong>Affected code:</strong></p>
<ul>
<li><code>api/src/services/messaging.js</code>
<ul>
<li><code>updateMessageTaskStates</code> (~line 281+)</li>
<li><code>getTaskForMessage</code> (~line 29)</li>
<li><code>getTaskAndDocForMessage</code> (~line 34)</li>
</ul>
</li>
</ul>
<p>Inside <code>updateMessageTaskStates</code>, rows are mapped without filtering missing docs:</p>
<pre><code class="language-js">const docs = results.rows.map(r =&gt; r.doc);
const stateChangesByDocId = applyTaskStateChangesToDocs(taskStateChanges, docs);
</code></pre>
<p>If any row has no <code>doc</code>, the call chain below crashes because <code>doc</code> is <code>undefined</code>:</p>
<pre><code>applyTaskStateChangesToDocs(...)
  → getTaskAndDocForMessage(...)
    → getTaskForMessage(messageId, doc)
      → getTaskFromMessage(doc.tasks, uuid)  // ❌ crashes here
</code></pre>
<blockquote>
<p>The function already handles partial failures for <code>bulkDocs</code> but does <strong>not</strong> defend against missing docs before task traversal.</p>
</blockquote>
<hr>
<h2>To Reproduce</h2>
<ol>
<li>Stub <code>db.medic.query('medic-sms/messages_by_uuid', ...)</code> to return rows with IDs.</li>
<li>Stub <code>db.medic.allDocs(...)</code> to return one row <strong>with</strong> <code>doc</code> and one row <strong>without</strong> <code>doc</code>.</li>
<li>Call <code>updateMessageTaskStates(...)</code>.</li>
<li>Observe crash before retry logic or graceful completion.</li>
</ol>
<hr>
<h2>Expected Behavior</h2>
<ul>
<li>Rows missing <code>doc</code> should be silently skipped.</li>
<li>State changes referencing unknown/missing docs should be logged and dropped.</li>
<li>The function should still update all valid docs in the same batch and return a stable <code>{ saved: n }</code> response.</li>
</ul>
<hr>
<h2>Actual Behavior</h2>
<p>A missing <code>doc</code> triggers an unhandled exception in task lookup, crashing the entire update operation.</p>
<hr>
<h2>Logs</h2>
<pre><code>TypeError: Cannot read properties of undefined (reading 'tasks')
    at getTaskFromMessage (api/src/services/messaging.js)
    at getTaskForMessage (api/src/services/messaging.js)
    at getTaskAndDocForMessage (api/src/services/messaging.js)
    at applyTaskStateChangesToDocs (api/src/services/messaging.js)
    at updateMessageTaskStates (api/src/services/messaging.js)
</code></pre>
<hr>
<h2>Screenshots</h2>
<p>N/A</p>
<hr>
<h2>Environment</h2>

Field | Value
-- | --
Instance | N/A (reproduced at unit test level)
Browser | N/A
Client platform | N/A
App | api
Version | N/A

<hr>
<h2>Additional Context</h2>
<p><code>allDocs</code> with <code>{ keys, include_docs: true }</code> does <strong>not</strong> guarantee a <code>doc</code> for every key — keys can be stale or deleted between queries. This is a known CouchDB/PouchDB behaviour, not an edge case.</p>
<h3>Proposed Fix (~55–70 lines across service + tests)</h3>
<ol>
<li>
<p><strong>Sanitize rows</strong> before processing in <code>updateMessageTaskStates</code>:</p>
<pre><code class="language-js">const docs = results.rows.map(r =&gt; r.doc).filter(Boolean);
</code></pre>
</li>
<li>
<p><strong>Add a defensive guard</strong> in <code>getTaskAndDocForMessage</code> to skip falsy docs early.</p>
</li>
<li>
<p><strong>Add logging</strong> for dropped state changes when no doc is available.</p>
</li>
<li>
<p><strong>Add unit tests</strong> in <code>api/tests/mocha/services/messaging.spec.js</code>:</p>
<ul>
<li>Should skip missing docs without throwing.</li>
<li>Should still update valid docs in the same batch.</li>
<li>Should return the correct <code>saved</code> count.</li>
</ul>
</li>
</ol>
<h3>Acceptance Criteria</h3>
<ul>
<li>[ ] No crash when <code>allDocs</code> includes rows without <code>doc</code>.</li>
<li>[ ] Existing successful update path is unchanged.</li>
<li>[ ] New tests pass and prove robustness for partial <code>allDocs</code> results.</li>
</ul></body></html>

### PR DESCRIPTION
**Pull request by Manishnemade12**
# Description

Fixes #10944.

`updateMessageTaskStates` could crash when `db.medic.allDocs({ include_docs: true })` returned rows without `doc` (for example `not_found` or deleted rows).

This PR:
- Adds defensive guards for falsy docs in message/task lookup helpers.
- Sanitizes `allDocs` rows before applying task state updates.
- Adds a warning log when missing docs are returned.
- Adds a regression unit test to verify missing-doc rows are ignored and valid updates still succeed.

Validation:
- Added/updated unit test in `api/tests/mocha/services/messaging.spec.js`.
- Static file checks pass.
- Local command `UNIT_TEST_ENV=1 npx mocha api/tests/mocha/services/messaging.spec.js` was attempted, but environment is missing `chai-exclude` in current install state.

# Code review checklist

- [ ] UI/UX backwards compatible: Test it works for the new design (enabled by default). And test it works in the old design, enable `can_view_old_navigation` permission to see the old design. Test it has appropriate design for RTL languages. 
- [x] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/)
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [x] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [x] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.
- [x] AI disclosure: Please disclose use of AI per [the guidelines](https://docs.communityhealthtoolkit.org/community/contributing/ai-guidelines/).

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
@Manishnemade12 can you fix the sonar linting issue, please?

**Comment by Manishnemade12** [comment]
@dianabarsan i fixed sonar linting issue , can you please review it now

File: api/src/services/messaging.js [hunk 1]
```diff
@@ -27,12 +27,18 @@ const getTaskFromMessage = (tasks, uuid) => {
 };
 
 const getTaskForMessage = (uuid, doc) => {
+  if (!doc) {
```
Comment (dianabarsan) [inline api/src/services/messaging.js]: The `.filter(Boolean)` you added in `updateMessageTaskStates` (line 300-302) already guarantees every element of `docs` passed downstream is a truthy document. Combined with the fact that `getTaskForMessage` and `getTaskAndDocForMessage` are module-private (neither is in `module.exports`), the only call chain into them is `updateMessageTaskStates` 
  → `applyTaskStateChangesToDocs` → `getTaskAndDocForMessage` → `getTaskForMessage`. The recursive retry at line 326 re-enters `updateMessageTaskStates` so the filter applies there too.                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                                                                
  That makes the two `if (!doc)` guards in `getTaskForMessage` (line 30) and `getTaskAndDocForMessage` (line 36) unreachable. Could you drop them? Filtering at the boundary is the right pattern; the inner guards are dead code that will only confuse future readers.

File: api/src/services/messaging.js [hunk 2]
```diff
@@ -292,7 +298,12 @@ module.exports = {
         return db.medic.allDocs({ keys: uniqueIds, include_docs: true });
       })
       .then(results => {
-        const docs = results.rows.map(r => r.doc);
+        const docs = results.rows
+          .map(r => r?.doc)
+          .filter(Boolean);
+        if (docs.length !== results.rows.length) {
```
Comment (dianabarsan) [inline api/src/services/messaging.js]: Please remove this comment. It's not actionable in any way and will just pollute logs.

**Review by dianabarsan — CHANGES_REQUESTED** [review]

File: api/src/services/messaging.js (hunk 1 above)
Comment (Manishnemade12) [reply api/src/services/messaging.js]: i Removed both inner !doc guards as suggested. Since updateMessageTaskStates already filters falsy docs at the boundary, those checks were unreachable and just added noise.

File: api/src/services/messaging.js (hunk 2 above)
Comment (Manishnemade12) [reply api/src/services/messaging.js]: Removed this warning log in messaging.js; agreed it wasn’t actionable and could create unnecessary noise in production logs.

**Comment by Manishnemade12** [comment]
@dianabarsan i addressed your all comments , can you please review it now

**Review by dianabarsan — APPROVED** [review]
⭐
